File: Leetcode 1701-1800/1722/1722.py
# Leetcode 1722: Minimize Hamming Distance After Swap Operations
# https://leetcode.com/problems/minimize-hamming-distance-after-swap-operations/
# Solved on 21st of April, 2026
from collections import defaultdict, Counter


class Solution:
    def minimumHammingDistance(self, source: list[int], target: list[int], allowedSwaps: list[list[int]]) -> int:
        """
        Calculates the minimum Hamming distance between source and target arrays after any number of allowed swaps.

        :param source: The initial list of integers.
        :param target: The target list of integers to compare against.
        :param allowedSwaps: A list of index pairs that can be swapped in the source array.
        :return: The minimum possible Hamming distance between the modified source and the target.
        """
        arrayLength = len(source)
        parentArray = list(range(arrayLength))
        rankArray = [1] * arrayLength

        def findRoot(nodeId):
            if parentArray[nodeId] != nodeId:
                parentArray[nodeId] = findRoot(parentArray[nodeId])
            return parentArray[nodeId]

        def unionNodes(nodeA, nodeB):
            rootA = findRoot(nodeA)
            rootB = findRoot(nodeB)
            if rootA != rootB:
                if rankArray[rootA] > rankArray[rootB]:
                    parentArray[rootB] = rootA
                elif rankArray[rootA] < rankArray[rootB]:
                    parentArray[rootA] = rootB
                else:
                    parentArray[rootB] = rootA
                    rankArray[rootA] += 1

        for swapPair in allowedSwaps:
            unionNodes(swapPair[0], swapPair[1])

        elementCountMap = defaultdict(Counter)

        for currIndex in range(arrayLength):
            rootIndex = findRoot(currIndex)
            elementCountMap[rootIndex][source[currIndex]] += 1

        minDistance = 0

        for currIndex in range(arrayLength):
            rootIndex = findRoot(currIndex)
            targetValue = target[currIndex]
            if elementCountMap[rootIndex][targetValue] > 0:
                elementCountMap[rootIndex][targetValue] -= 1
            else:
                minDistance += 1

        return minDistance

File: Leetcode 1701-1800/1722/1722.java
// Leetcode 1722: Minimize Hamming Distance After Swap Operations
// https://leetcode.com/problems/minimize-hamming-distance-after-swap-operations/
// Solved on 21st of April, 2026
import java.util.HashMap;
import java.util.Map;

class Solution {
    /**
     * Calculates the minimum Hamming distance between source and target arrays after any number of allowed swaps.
     *
     * @param source       The initial array of integers.
     * @param target       The target array of integers to compare against.
     * @param allowedSwaps An array of pairs representing indices that can be swapped.
     * @return The minimum possible Hamming distance between the modified source and the target.
     */
    public int minimumHammingDistance(int[] source, int[] target, int[][] allowedSwaps) {
        int n = source.length;
        int[] parent = new int[n];
        for (int i = 0; i < n; i++) {
            parent[i] = i;
        }
        for (int[] swap : allowedSwaps) {
            int rootA = find(parent, swap[0]);
            int rootB = find(parent, swap[1]);
            if (rootA != rootB) {
                parent[rootA] = rootB;
            }
        }
        Map<Integer, Integer>[] freqMaps = new HashMap[n];
        for (int i = 0; i < n; i++) {
            int root = find(parent, i);
            if (freqMaps[root] == null) {
                freqMaps[root] = new HashMap<>();
            }
            freqMaps[root].put(source[i], freqMaps[root].getOrDefault(source[i], 0) + 1);
        }
        int matches = 0;
        for (int i = 0; i < n; i++) {
            int root = find(parent, i);
            if (freqMaps[root] != null && freqMaps[root].getOrDefault(target[i], 0) > 0) {
                matches++;
                freqMaps[root].put(target[i], freqMaps[root].get(target[i]) - 1);
            }
        }
        return n - matches;
    }

    private int find(int[] parent, int i) {
        if (parent[i] == i) {
            return i;
        }
        return parent[i] = find(parent, parent[i]);
    }
}